?Sized and Debug well-known traits are incompatible

Consider

```
type str {
    wellknown_traits(?Sized,Debug);
}
```

This produces the following rust code:

```rust
#[allow(non_snake_case)]
#[no_mangle]
pub extern "C" fn __zngur_str_debug_pretty_e11(v: *mut u8) {
    eprintln!("{:#?}", unsafe { &*(v as *mut str) });
}

#[allow(non_snake_case)]
#[no_mangle]
pub extern "C" fn __zngur_str_debug_print_e11(v: *mut u8) {
    eprintln!("{:?}", unsafe { &*(v as *mut str) });
}
```

which produces this error when compiled,

```
error[E0607]: cannot cast thin pointer `*mut u8` to wide pointer `*mut str`
  --> src/generated.rs:46:35
   |
46 |     eprintln!("{:#?}", unsafe { &*(v as *mut str) });
   |                                   ^^^^^^^^^^^^^^^

error[E0607]: cannot cast thin pointer `*mut u8` to wide pointer `*mut str`
  --> src/generated.rs:52:34
   |
52 |     eprintln!("{:?}", unsafe { &*(v as *mut str) });
   |                                  ^^^^^^^^^^^^^^^
```

--- a/examples/regression_test1/.gitignore
+++ b/examples/regression_test1/.gitignore
@@ -1,0 +1,3 @@
+generated.h
+generated.rs
+generated.cpp

--- a/examples/regression_test1/Cargo.toml
+++ b/examples/regression_test1/Cargo.toml
@@ -1,0 +1,13 @@
+[package]
+name = "example-regression-test1"
+version = "0.5.0"
+edition.workspace = true
+license.workspace = true
+publish = false
+
+[lib]
+crate-type = ["staticlib"]
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]

--- a/examples/regression_test1/Makefile
+++ b/examples/regression_test1/Makefile
@@ -1,0 +1,10 @@
+a.out: main.cpp generated.h src/generated.rs src/lib.rs ../../target/release/libexample_regression_test1.a
+	${CXX} main.cpp -g -L ../../target/release/ -l example_regression_test1
+
+../../target/release/libexample_regression_test1.a: FORCE
+	cargo build --release
+
+generated.h ./src/generated.rs: main.zng
+	cd ../../zngur-cli && cargo run g ../examples/regression_test1/main.zng
+
+FORCE: ;

--- a/examples/regression_test1/README.md
+++ b/examples/regression_test1/README.md
@@ -1,0 +1,10 @@
+# Example: Regression test 1
+
+A example, used to check previous Zngur problems in CI.
+
+To run this example:
+
+```
+make
+./a.out
+```

--- a/examples/regression_test1/main.cpp
+++ b/examples/regression_test1/main.cpp
@@ -1,0 +1,20 @@
+#include <iostream>
+#include <vector>
+
+#include "./generated.h"
+
+int main() {
+  // Test dbg works for Ref and RefMut
+  rust::Ref<rust::Str> v1 = rust::Str::from_char_star("foo");
+  zngur_dbg(v1);
+  rust::std::string::String v2 = v1.to_owned();
+  zngur_dbg(v2);
+  rust::Ref<rust::std::string::String> v3 = v2;
+  zngur_dbg(v3);
+  rust::std::string::String v4 = std::move(zngur_dbg(v2));
+  zngur_dbg(v4);
+  rust::RefMut<rust::std::string::String> v5 = v4;
+  zngur_dbg(v5);
+  v5.push_str(zngur_dbg(rust::Str::from_char_star("bar")));
+  zngur_dbg(v4);
+}

--- a/examples/regression_test1/main.zng
+++ b/examples/regression_test1/main.zng
@@ -1,0 +1,16 @@
+#convert_panic_to_exception
+
+type str {
+    wellknown_traits(?Sized, Debug);
+
+    fn as_ptr(&self) -> *const u8;
+    fn len(&self) -> usize;
+    fn to_owned(&self) -> ::std::string::String;
+}
+
+type ::std::string::String {
+    #layout(size = 24, align = 8);
+    wellknown_traits(Debug);
+
+    fn push_str(&mut self, &str);
+}

--- a/examples/regression_test1/src/lib.rs
+++ b/examples/regression_test1/src/lib.rs
@@ -1,0 +1,1 @@
+mod generated;

--- a/zngur-generator/src/cpp.rs
+++ b/zngur-generator/src/cpp.rs
@@ -517,6 +517,7 @@ struct {ref_kind}< {ty} > {{
 private:
     ::std::array<size_t, 2> data;
     friend uint8_t* ::rust::__zngur_internal_data_ptr< ::rust::{ref_kind}< {ty} > >(const ::rust::{ref_kind}< {ty} >& t);
+    friend void ::rust::zngur_pretty_print< ::rust::{ref_kind}< {ty} > >(::rust::{ref_kind}< {ty} > const& t);
 "#,
                     ty = self.ty,
                 )?;
@@ -537,6 +538,7 @@ struct {ref_kind}< {ty} > {{
 private:
     size_t data;
     friend uint8_t* ::rust::__zngur_internal_data_ptr< ::rust::{ref_kind}< {ty} > >(const ::rust::{ref_kind}< {ty} >& t);
+    friend void ::rust::zngur_pretty_print< ::rust::{ref_kind}< {ty} > >(::rust::{ref_kind}< {ty} > const& t);
 "#,
                     ty = self.ty,
                 )?;
@@ -1254,24 +1256,60 @@ auto data_as_impl = &args;
                 )?;
             }
         }
+        let is_unsized = self
+            .wellknown_traits
+            .contains(&ZngurWellknownTraitData::Unsized);
         for tr in &self.wellknown_traits {
             match tr {
                 ZngurWellknownTraitData::Debug {
                     pretty_print,
                     debug_print: _, // TODO: use it
                 } => {
-                    writeln!(
-                        state,
-                        r#"
+                    if !is_unsized {
+                        writeln!(
+                            state,
+                            r#"
             namespace rust {{
                 template<>
                 inline void zngur_pretty_print< {ty} >({ty} const& t) {{
                     ::rust::__zngur_internal_check_init< {ty} >(t);
                     {pretty_print}(&t.data[0]);
                 }}
+
+                template<>
+                inline void zngur_pretty_print< Ref< {ty} > >(Ref< {ty} > const& t) {{
+                    ::rust::__zngur_internal_check_init< Ref< {ty} > >(t);
+                    {pretty_print}(reinterpret_cast<uint8_t*>(t.data));
+                }}
+
+                template<>
+                inline void zngur_pretty_print< RefMut< {ty} > >(RefMut< {ty} > const& t) {{
+                    ::rust::__zngur_internal_check_init< RefMut< {ty} > >(t);
+                    {pretty_print}(reinterpret_cast<uint8_t*>(t.data));
+                }}
             }}"#,
-                        ty = self.ty,
-                    )?;
+                            ty = self.ty,
+                        )?;
+                    } else {
+                        writeln!(
+                            state,
+                            r#"
+            namespace rust {{
+                template<>
+                inline void zngur_pretty_print< Ref< {ty} > >(Ref< {ty} > const& t) {{
+                    ::rust::__zngur_internal_check_init< Ref< {ty} > >(t);
+                    {pretty_print}(::rust::__zngur_internal_data_ptr< Ref< {ty} > >(t));
+                }}
+
+                template<>
+                inline void zngur_pretty_print< RefMut< {ty} > >(RefMut< {ty} > const& t) {{
+                    ::rust::__zngur_internal_check_init< RefMut< {ty} > >(t);
+                    {pretty_print}(::rust::__zngur_internal_data_ptr< RefMut< {ty} > >(t));
+                }}
+            }}"#,
+                            ty = self.ty,
+                        )?;
+                    }
                 }
                 ZngurWellknownTraitData::Unsized
                 | ZngurWellknownTraitData::Copy
@@ -1516,6 +1554,7 @@ namespace rust {
         private:
             size_t data;
         friend uint8_t* ::rust::__zngur_internal_data_ptr<Ref< {ty} > >(const ::rust::Ref< {ty} >& t);
+        friend void ::rust::zngur_pretty_print< Ref< {ty} > >(Ref< {ty} > const& t);
     }};
 
     template<>
@@ -1533,6 +1572,7 @@ namespace rust {
         private:
             size_t data;
         friend uint8_t* ::rust::__zngur_internal_data_ptr<RefMut< {ty} > >(const ::rust::RefMut< {ty} >& t);
+        friend void ::rust::zngur_pretty_print< Ref< {ty} > >(Ref< {ty} > const& t);
     }};
 "#
             )?;

--- a/zngur-generator/src/lib.rs
+++ b/zngur-generator/src/lib.rs
@@ -118,8 +118,11 @@ impl ZngurGenerator {
                     });
                 }
             }
+            let is_unsized = ty_def
+                .wellknown_traits
+                .contains(&ZngurWellknownTrait::Unsized);
             for wellknown_trait in ty_def.wellknown_traits {
-                let data = rust_file.add_wellknown_trait(&ty_def.ty, wellknown_trait);
+                let data = rust_file.add_wellknown_trait(&ty_def.ty, wellknown_trait, is_unsized);
                 wellknown_traits.push(data);
             }
             for method_details in ty_def.methods {

--- a/zngur-generator/src/rust.rs
+++ b/zngur-generator/src/rust.rs
@@ -706,6 +706,7 @@ pub extern "C" fn {mangled_name}("#
         &mut self,
         ty: &RustType,
         wellknown_trait: ZngurWellknownTrait,
+        is_unsized: bool,
     ) -> ZngurWellknownTraitData {
         match wellknown_trait {
             ZngurWellknownTrait::Unsized => ZngurWellknownTraitData::Unsized,
@@ -726,13 +727,18 @@ pub extern "C" fn {drop_in_place}(v: *mut u8) {{ unsafe {{
             ZngurWellknownTrait::Debug => {
                 let pretty_print = mangle_name(&format!("{ty}=debug_pretty"));
                 let debug_print = mangle_name(&format!("{ty}=debug_print"));
+                let dbg_ty = if !is_unsized {
+                    format!("{ty}")
+                } else {
+                    format!("&{ty}")
+                };
                 wln!(
                     self,
                     r#"
 #[allow(non_snake_case)]
 #[no_mangle]
 pub extern "C" fn {pretty_print}(v: *mut u8) {{
-    eprintln!("{{:#?}}", unsafe {{ &*(v as *mut {ty}) }});
+    eprintln!("{{:#?}}", unsafe {{ &*(v as *mut {dbg_ty}) }});
 }}"#
                 );
                 wln!(
@@ -741,7 +747,7 @@ pub extern "C" fn {pretty_print}(v: *mut u8) {{
 #[allow(non_snake_case)]
 #[no_mangle]
 pub extern "C" fn {debug_print}(v: *mut u8) {{
-    eprintln!("{{:?}}", unsafe {{ &*(v as *mut {ty}) }});
+    eprintln!("{{:?}}", unsafe {{ &*(v as *mut {dbg_ty}) }});
 }}"#
                 );
                 ZngurWellknownTraitData::Debug {

